### PR TITLE
jbibla+fedekunze/1501 Modal Vote

### LINKED ISSUE #1501: Voting
Allow users to vote on proposals.

### PR DESCRIPTION
**Pull request by fedekunze**
Depends #1531 
Closes #1501

_Description:_

TODO:

- [x] Modal to vote based on [wireframe](https://app.zeplin.io/project/59e66beff6721ea8bf562e5b/screen/5bc0f387a55ed15fd3fda054) design
~~Link to explain the types of proposals~~
- [x] Disable other option buttons if an option is selected and select the clicked button
- [x] Styling cc: @jbibla
- [x] Integrate modal with PageProposal
- [ ] E2E tests #1550 

❤️ Thank you!

---

* [x] Added entries in `CHANGELOG.md` with issue # and GitHub username
* [ ] Reviewed `Files changed` in the github PR explorer

### CODE REVIEW

**Comment by fedekunze** [comment]
> this is governance 2 kind of stuff 😸

True ! I will add it to the scope of the gov2 issue

File: app/src/renderer/components/governance/ModalVote.vue [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    TmFormMsg
+  },
+  data: () => ({
+    option: ``
+  }),
+  validations() {
+    return {
+      option: {
+        required,
+        isValid
```
Comment (faboweb) [inline app/src/renderer/components/governance/ModalVote.vue]: In which case would this be invalid?

File: app/src/renderer/connectors/lcdClientMock.js [hunk 2]
```diff
@@ -21,6 +21,9 @@ let state = {
       address: addresses[0]
     }
   ],
+  wallet: {
```
[the remark this answers is not in the document]
Comment (fedekunze) [reply app/src/renderer/connectors/lcdClientMock.js]: Oh that was for the issue that I couldn't retrieve the state, but @NodeGuy fixed it so I can delete it now

File: app/src/renderer/connectors/lcdClientMock.js [hunk 3]
```diff
@@ -387,6 +417,7 @@ let state = {
         option: `no_with_veto`
       }
     ],
+    2: [],
```
Comment (faboweb) [inline app/src/renderer/connectors/lcdClientMock.js]: do the properties reflect the propsal_ids? should they then maybe be strings?

Comment (fedekunze) [reply app/src/renderer/connectors/lcdClientMock.js]: good point, I'll try switching them to strings

File: app/src/renderer/connectors/lcdClientMock.js [hunk 4]
```diff
@@ -865,8 +897,67 @@ module.exports = {
...
+    let results = []
+    let fromKey = state.keys.find(a => a.name === name)
+    let fromAccount = state.accounts[fromKey.address]
+
+    if (fromAccount == null) {
+      results.push(txResult(1, `Nonexistent account`))
+      return results
+    }
+    // check nonce
+    if (parseInt(fromAccount.sequence) !== parseInt(sequence)) {
```
Comment (faboweb) [inline app/src/renderer/connectors/lcdClientMock.js]: just an idea: we can probably refactor all this nonce and account checking. but not in the scope of this PR.

Comment (fedekunze) [reply app/src/renderer/connectors/lcdClientMock.js]: Cool, can u open an issue with the details ?

File: test/unit/specs/components/governance/PageProposal.spec.js [hunk 5]
```diff
@@ -10,43 +15,153 @@ let status = {
...
-  let wrapper
-  let { mount } = setup()
-
-  beforeEach(() => {
-    let instance = mount(PageProposal, {
-      propsData: {
-        proposal,
-        status
-      }
+  describe(`with magic`, () => {
```
Comment (faboweb) [inline test/unit/specs/components/governance/PageProposal.spec.js]: lol

File: app/src/renderer/connectors/lcdClientMock.js (hunk 3 above)
Comment (fedekunze) [reply app/src/renderer/connectors/lcdClientMock.js]: keys can't be strings. I get a `unexpected token ...` error

File: app/src/renderer/components/governance/ModalVote.vue (hunk 1 above)
Comment (faboweb) [reply app/src/renderer/components/governance/ModalVote.vue]: correct me if I am wrong, but the user is only able to produce the states `yes` , `no`, `no_with_veto`, `abstain` and ``. so `isValid` equals `required`.

File: app/src/renderer/components/governance/ModalVote.vue [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+      }
+    }
+  },
+  methods: {
+    close() {
+      this.$emit(`update:showModalVote`, false)
+    },
+    voteYes() {
```
[the remark this answers is not in the document]
Comment (faboweb) [reply app/src/renderer/components/governance/ModalVote.vue]: yes I think so

File: test/unit/specs/components/governance/PageProposal.spec.js [hunk 7]
```diff
@@ -10,43 +15,153 @@ let status = {
...
+          status
+        }
+      })
+      wrapper.update()
+      expect(wrapper.vm.voteBlock).toBe(`the same block`)
+    })
+
+    describe(`Modal onVote`, () => {
+      it(`enables voting if the proposal is Active`, () => {
+        let status = { button: `vote` }
```
Comment (faboweb) [inline test/unit/specs/components/governance/PageProposal.spec.js]: just thinking: status seems to be a complex object that doesn't seem self explanatory. Maybe we could split the object in a refactor.

File: test/unit/specs/components/governance/PageProposal.spec.js [hunk 8]
```diff
@@ -10,43 +15,153 @@ let status = {
...
   })
 
-  it(`should return the end of the sentence`, () => {
-    proposal.submit_block = `135`
-    let { wrapper } = mount(PageProposal, {
-      propsData: {
-        proposal,
-        status
-      }
+  describe(`without magic`, () => {
```
Comment (faboweb) [inline test/unit/specs/components/governance/PageProposal.spec.js]: cool, you got this. :)

File: test/unit/specs/lcdClientMock.spec.js [hunk 9]
```diff
@@ -850,52 +850,172 @@ describe(`LCD Client Mock`, () => {
...
-    let proposals = lcdClientMock.state.proposals
-    let proposalsRes = await client.getProposals()
-    expect(proposalsRes).toEqual(proposals)
-  })
+  describe(`Governance`, () => {
+    describe(`Proposals`, () => {
+      it(`fetches all governance proposals`, async () => {
+        let proposals = lcdClientMock.state.proposals
+        let proposalsRes = await client.getProposals()
+        expect(proposalsRes).toBeDefined()
```
Comment (faboweb) [inline test/unit/specs/lcdClientMock.spec.js]: cool safety check

File: test/unit/specs/components/governance/PageProposal.spec.js (hunk 7 above)
Comment (fedekunze) [reply test/unit/specs/components/governance/PageProposal.spec.js]: cc: @jbibla ^^

File: test/unit/specs/store/governance/votes.spec.js [hunk 10]
```diff
@@ -12,6 +12,7 @@ describe(`Module: Votes`, () => {
 
   it(`adds votes to state`, () => {
     let { mutations, state } = module
+    console.log(proposals)
```
Comment (faboweb) [inline test/unit/specs/store/governance/votes.spec.js]: remove?

File: app/src/renderer/connectors/lcdClientMock.js (hunk 3 above)
Comment (faboweb) [reply app/src/renderer/connectors/lcdClientMock.js]: Did you use backticks or quotemarks? Keys are not allowed to have backticks.

File: app/src/renderer/components/governance/ModalVote.vue (hunk 6 above)
[the remark this answers is not in the document]
Comment (fedekunze) [reply app/src/renderer/components/governance/ModalVote.vue]: done

File: app/src/renderer/connectors/lcdClientMock.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (fedekunze) [reply app/src/renderer/connectors/lcdClientMock.js]: done

File: app/src/renderer/connectors/lcdClientMock.js (hunk 3 above)
Comment (fedekunze) [reply app/src/renderer/connectors/lcdClientMock.js]: done

File: test/unit/specs/store/governance/votes.spec.js (hunk 10 above)
Comment (fedekunze) [reply test/unit/specs/store/governance/votes.spec.js]: done

File: app/src/renderer/connectors/lcdClientMock.js (hunk 4 above)
Comment (faboweb) [reply app/src/renderer/connectors/lcdClientMock.js]: https://github.com/cosmos/voyager/issues/1551

File: app/src/renderer/components/governance/ModalVote.vue (hunk 1 above)
Comment (fedekunze) [reply app/src/renderer/components/governance/ModalVote.vue]: correct, the button to vote is only disabled if in this case the user hasn't selected an option

Comment (fedekunze) [reply app/src/renderer/components/governance/ModalVote.vue]: I just added it as a doublecheck ;)

**Pull request merged**